docs/src/pages/components/buttons/CustomizedButtons.js

```diff
@@ -66,6 +66,66 @@ const theme = createMuiTheme({
   palette: {
     primary: green,
   },
+  additions: {
+    MuiButton: {
+      dashed: {
+        padding: '5px 15px',
+        border: '5px dashed red',
+        '&$disabled': {
+          border: `5px dashed red`,
+        },
+      },
+      tertiery: {
+        backgroundColor: 'yellow',
+        padding: 20,
+      },
+    },
+  },
+  variants: {
+    MuiButton: [
+      {
+        trigger: (props) => props.variant === 'red',
+        styles: {
+          color: 'red'
+        }
+      },
+      {
+        trigger: (props) => props.size === 'xxsmall',
+        styles: {
+          fontSize: '5px'
+        }
+      },
+      {
+        trigger: (props) => props.toggle,
+        styles: {
+          background: 'green'
+        }
+      },
+      {
+        trigger: (props) => props.toggle === 'red',
+        styles: {
+          backgroundColor: 'red'
+        }
+      }
+    ],
+  },
+  variantsV2: {
+    MuiButton: [
+      {
+        matcher: { variant: 'dashed', color: 'yellow' },
```

Comment:
We could use a syntax closed to styled system:

```
  variant({
    prop: 'size',
    variants: {
      big: {
        fontSize: 4,
        lineHeight: 'heading',
      },
      small: {
        fontSize: 1,
        lineHeight: 'body',
      },
    }
  })
```

but I am not sure how to use combination of props with that syntax, so I went with the proposed one...

Reply:
We can ask ourselves if the combination of props is this a frequent requirement? It comes from time to time, for instance:

https://github.com/mui-org/material-ui/blob/6ae4a760d8114fce952780b1805516724abbf9b4/packages/material-ui/src/Button/Button.js#L185

It seems that we need to support it, but we could offer a different tradeoff for it. For instance, we might want to encourage direct matchers.

Reply:
Maybe we can have two alternatives for defining them:

```
theme.variants = {
  MuiButton: [
    // single prop
    {
      prop: 'variant',
      value: 'red',
      styles: {
        color: 'red',
      }
    },
    // combination of props
    {
      matcher: { toggle: true, variant: 'red' },
      styles: {
        background: 'red',
      }
    }
    // other alternative prop
    {
      matcher: [{prop: 'variant',  value: 'red'}, {prop: 'toggle', value: true}],
      styles: {
        color: 'red',
      }
    },
  ],
}
```

But to be honest, I like the best the matcher syntax, as the same one can be used for both use-cases, and it is closest to how we think about the props matcher in js anyway...

Reply:
> I like the best the matcher syntax

Sounds great